### DISCUSSION TITLE
Querying workspace dimensions and viewport position.

### DISCUSSION THREAD

**Discussion by olistrik** [YaLTeR/niri#500] [Q&A]
Hi there, I'm currently trying out Niri as I love the idea of not being constrained to the width of my tiny laptop screen.

I was wondering if it's possible to query the state of a workspace and the viewport, something like `swaymsg get_tree` perhaps?

I've noticed that I've started to loose track of where exactly I am in a workspace, and had the idea to add something like a scrollbar or a minimap within my ags bar to give myself a visual overview.

**Comment by olistrik** [comment]
I'm guessing this is related to https://github.com/YaLTeR/niri/pull/294

**Reply by YaLTeR** [reply]
Hey, not really, that's unrelated. It's more related to #453 (and adding events like this). However, I'm not sure adding continuous data like this to IPC is a good idea because it can change very frequently (e.g. during a mouse gesture with a 1000 Hz mouse, 1000 times per second).

**Reply by olistrik** [reply]
Ah, thanks for moving this to discussions, I didn't realize those were a thing! I've had a skim now through the other discussions, and I think this is also related to https://github.com/YaLTeR/niri/discussions/352 as that is the kind of thing I'm trying to produce. Though through a 3rd party widget over IPC rather than as a builtin feature.

I think it's still worth keeping this discussion open, as I can think of other things I might want to implement later with this kind of functionality.

> I'm not sure adding continuous data like this to IPC is a good idea because it can change very frequently

I hadn't considered this, honestly I was just planning on polling for the state every few seconds or so initially like I would for waybar extensions, but that's going to result in a very laggy ux. It's also potentially quite a lot of unnecessary overhead if every second it's asking for a json tree that hasn't changed. Doing it through callbacks or event emitters would be a much cleaner approach if it can be implemented correctly. 

Would it be acceptable to debounce or throttle continuous data? Or alternatively only emit the continuous data as a result of discrete events?

In javascript binding to drag events can also be problematic for similar reasons so typically unless it's absolutely necessary, I bind to `dragstart`,  `dragend`, etc and only perform heavy logic when those are fired. I dunno if that would be possible for Niri as I haven't looked into the source yet.

For what I had in mind for the minimap / scrollbar, I could probably achieve a good UX if I could at a minimum listen to:

- `workspaceResized` which emits the new width/height once a resize is completed (so not during mouse resizing)
-  `windowFocused` which emits amongst other things, the bounds of the window at that moment.

Ultimately, if some events are not possible or impractical, I can achieve the same effect if I could manually request the workspace and focused window bounds and do so when more specific events occur that would result in a change in the workspace size. 

Exactly what kind of interface is possible or that you'd like to export I'm not sure.

**Reply by YaLTeR** [reply]
> Would it be acceptable to debounce or throttle continuous data?

Perhaps; something like when the animations are done and there are no ongoing gestures.

> windowFocused which emits amongst other things, the bounds of the window at that moment.

Do you really need the focused window position?

**Reply by olistrik** [reply]
> Do you really need the focused window position?

It depends really on what events could be subscribed to and what data I could retrieve. Ultimately I'm either looking for the position and width of the viewport or for the focused column. I might be wrong, but from what I understand it's not possible to create columns within columns, so the horizontal position and width of a focused window should be equivalent to the column that contains it right? 

Either way, I can give an example of what I'm thinking of doing.

Say we have a workspace with four columns, each sized proportionally to the screen width with the ratios 1.0, 0.5, 0.5, and 1.0 respectively. This would mean that the workspace is 3x the width of the viewport.

There's a number of ways that this could be presented but the three ideas I had in mind could look like this:

![image](https://github.com/YaLTeR/niri/assets/17579507/0fe67487-af6c-4767-a8d7-9b2850ca6f09)

In all cases the gray bar represents the total width of the workspace, and the 3rd column is currently holding the focus, and is positioned on the left side of the viewport.

In the first example the bar is continuous and unbroken and the orange segment represents the proportion of workspace that currently visible in the viewport. In order to keep the scale and position of the viewport indicator accurate, I'd need to know:
  
  - When the workspace dimensions change. 
  - When the viewport moves.
  - When the output changes.
  - Optionally, when the workspace is focused or unfocused. 

The second example is similar to the first, except that the orange segment indicates the relative position and width of the focused column in the workspace. It's kinda a halfway house between the first and third examples, for this I'd need events about the focused window/column rather than the viewport and output.

The third I'd expect to be the most complicated as rather than being continuous, the gray bar representing the whole workspace is broken into segments that represent the relative width of each column within the workspace. The segment representing the currently focused column is then highlighted in orange. Unlike the first idea this option wouldn't require knowing either the workspace or the viewport dimensions, but would instead need to be notified of any changes to the dimensions of any of the columns within the workspace. Realistically speaking though I would expect an array of column widths would be a property available on `workspaceResized` events rather than having to subscribe to each column individually.

Personally of the three examples, my preference would be with 3 as it provides the most natural overview of both your position in the workspace and the structure of the workspace as a whole. Otherwise I think 1 is clearest behavior wise as it's pretty much identical to how all scrollbars work. 2 is the most confusing, as without the segmentation of 3 it's not immediately apparent if the orange section represents the viewport or the column and I think the natural assumption would be the viewport.

All of this is of course hypothetical, I haven't done a lot of work in rust, and nothing at all with wayland or smithay, so I don't know how much flexibility you have with event listeners and the like. I'm also sure you have your own ideas about the direction of IPC.

**Reply by YaLTeR** [reply]
> it's not possible to create columns within columns, so the horizontal position and width of a focused window should be equivalent to the column that contains it right?

It's not possible to create columns within columns, but the horizontal position and width of a window is also not equivalent to the containing column, because you can have a window with a large minimum width that it doesn't signal to the compositor, and a smaller window, in the same column (or just two differently-sized unresizable dialogs). So you'd need column positions and widths (rather than particular window position and width).

> There's a number of ways that this could be presented but the three ideas I had in mind could look like this:

I see, thanks for the examples. I'll keep them in mind when working on IPC.

**Comment by vredesbyyrd** [comment]
@olistrik Did you ever end up implementing this? I have also been thinking that something like a 'column indicator', or someway to indicate the users horizontal location within a workspace makes perfect sense for a scrolling wm. 

With more than a few columns open I often forget if a particular window is to the left or right, or, if there are any more windows to the left/right. 

The way I am currently handling this is with left/right struts, so the next/prev columns overflow into the active column. It works pretty well, my only gripe is when you are at the start or end of a workspace the struts are kind of a waste of space and look odd. I'm contemplating trying to hack together some sort of 'dynamic' struts with the ipc - If at start of workspace left strut = 0, if at the end of workspace right strut = 0. Not very pretty but it may work.

**Reply by olistrik** [reply]
@vredesbyyrd Not yet, unless something changed recently, there currently isn't a way over IPC to determine any information about columns.

I've not given up on the idea, it's just in limbo until either the functionality gets added to IPC or I find the time to add it myself.

**Reply by YaLTeR** [reply]
@calops is working on the IPC part

**Reply by olistrik** [reply]
@YaLTeR awesome! Is there somewhere tracking progress on that? I'm sure I'll read it in the release notes at some point, but it would be nice to watch a PR or something for notifications.

**Reply by YaLTeR** [reply]
No PR yet
